shell/platform/common/cpp/flutter_accessibility.h

```diff
@@ -0,0 +1,79 @@
+/// The accessibility node to be used in accessibility bridge. This class is
+/// responsible for providing native accessibility object with appropriate
+/// information, such as accessibility label/value/bounds.
+///
+/// While most methods have default implementations and are ready to be used
+/// as-is, the subclasses must override the GetNativeViewAccessible to return
+/// native accessibility objects. To do that, subclasses should create and
+/// maintain AXPlatformNode[s] which delegate their accessibility attributes to
+/// this class.
+///
+/// For desktop platforms, subclasses also need to override the GetBoundsRect
+/// to apply window-to-screen transform.
+///
+/// Lastly, each platform needs to implement the FlutterAccessibility::Create
+/// static method to inject its sublcass into accessibility bridge.
+class FlutterAccessibility : public AXPlatformNodeDelegateBase {
+ public:
+  //------------------------------------------------------------------------------
+  /// @brief      Creates a platform specific FlutterAccessibility. Ownership
+  ///             passes to the caller. This method will be called by
+  ///             accessibility bridge when it creates accessibility node. Each
+  ///             platform needs to implement this method in order to inject its
+  ///             subclass into the accessibility bridge.
+  static FlutterAccessibility* Create();
+
+  FlutterAccessibility();
+  ~FlutterAccessibility() override;
+
+  //------------------------------------------------------------------------------
+  /// @brief      Gets the accessibility bridge to which this accessibility node
+  ///             belongs.
+  AccessibilityBridge* GetBridge() const;
+
+  //------------------------------------------------------------------------------
+  /// @brief      Gets the underlying ax node for this accessibility node.
+  AXNode* GetAXNode() const;
+
+  // AXPlatformNodeDelegateBase override;
+  const AXNodeData& GetData() const override;
+  bool AccessibilityPerformAction(const AXActionData& data) override;
+  gfx::NativeViewAccessible GetParent() override;
+  gfx::NativeViewAccessible GetFocus() override;
+  int GetChildCount() const override;
+  gfx::NativeViewAccessible ChildAtIndex(int index) override;
+  gfx::Rect GetBoundsRect(const AXCoordinateSystem coordinate_system,
+                          const AXClippingBehavior clipping_behavior,
+                          AXOffscreenResult* offscreen_result) const override;
+  //------------------------------------------------------------------------------
+  /// @brief      Called only once, immediately after construction. The
+  ///             constructor doesn't take any arguments because in the Windows
+  ///             subclass we use a special function to construct a COM object.
+  ///             Subclasses must call super.
+  virtual void Init(AccessibilityBridge* bridge, AXNode* node);
+
+ private:
+  AXNode* ax_node_;
+  AccessibilityBridge* bridge_;
```

Comment:
Avoid raw pointers.

I'm also still questioning why we'd need a pointer from this object to an a11y bridge, but I might be misunderstanding what this object does.

Reply:
this object wraps AXNode and uses it to implement AXPlatformNodeDelegate API, it needs reference so that it can have access to the parent and its sibling which is required to implement the AXPlatformNodeDelegate API. Alternatively, we can store all these information in this node, but we will have to keep updating them when the structure of AXTree changes.